### ISSUE TITLE
Action Plan

### ISSUE THREAD

**Issue by FyZyX** [lucasmlofaro/tig#2]
## Concept
In order to apply a commit we need to know which snippets to affect (see #1 ). To do this, we must take the original commit message and generate a change plan from it; this change plan should consist of the atomic snippet operations (insertion, deletion, and replacement). In other words, every action should boil down to one of these essential actions.

## Methodology
We will use a large language model (LLM) to create the initial change plan. This is obviously challenging for a number of reasons, but at this stage of development, we will assume that such models are capable of presenting at least the top-level objectives, even if they may not understand the implications of making those changes. We will take more precautions to handle such consequences in subsequent procedures. The challenge here is that we don't know what changes the LLM will choose to make, so we can't provide specialized context like dependency or usage graphs at this stage (because we would have to do so for the entire codebase, which would be too much data).

## Providing Context
The most obvious things we can provide in the planning stage are:
- Project description/purpose
- Directory structure
- Relevant snippets
- Relevant documentation

### Project Description
We can allow users to specify a purpose via configuration. It's also possible that we could just grab the first chunk of the `README`.

### Directory Structure
We can introspect the directory structure by walking the file system from the root of the repository. Depending how big that gets, we may want to stop at a certain depth, but ideally we'd give as much context as possible here.

### Relevant Snippets
A semantic search on the snippets might provide some valuable context clues. We could provide 3-5 snippets depending how big they are and how relevant they appear based on the similarity score. Each snippet should also have useful metadata including a natural language description of the snippet.

### Relevant Documentation
If we chose to break up the documentation in chunks (assuming documentation exists), then we could also do a semantic search through that as well.

## Prompting
Once we've compiled all of that context, we can then write a prompt to ask for a change plan that only uses the atomic snippet operations. This is definitely more nuanced than it sounds, one example being adding or removing files and directories, but for now we can assume those can be abstracted over. Ideally, we should try to get back some kind of programmatic output that is easily parable. In theory, this output would tell us which files we will need to affect at minimum, and it may be necessary to do additional planning step for each file the LLM identifies as relevant.

We will also need a way to map each step to the relevant snippet, which will likely involve a search on the metadata associated with each snippet. For example, we are told to replace the snippet containing a specific function, and we have function names stored as metadata for each snippet, so we need to find the snippet corresponding to that function name.

## Executing the Plan
For each snippet that needs to be inserted or replaced, we must again prompt the LLM to update the code and consider potential downstream repercussions of making that change.

**Comment by seanscofield** [comment]
Looks pretty good so far!

Would the 3-5 snippets that we feed the LLM for context be fetched using the vector database that we're using? And do you think the vector database would be able to find a relevant code snippet if that snippet were a method within a class? I may have asked this the other day, but I'm ultimately wondering if the dot product operations performed by that database would be able to find "method FOO within class BAR" if the snippet for **method FOO** doesn't actually have a reference to **class BAR** in it (although I realize that information will be in the metadata).

**Comment by FyZyX** [comment]
@seanscofield yes, any similarity search will require a call to a vector database.
As for the method question, my instinct says that it would indeed locate the snippet simply because it would contain the name `foo`. Even that one word should dramatically increase the similarity of the snippet and commit message embeddings. If that turns out not to be the case, we could potentially append some keywords before we create the embedding of the commit message so that it's more inclined to match function declarations, but no need to go there unless the simpler method fails.
